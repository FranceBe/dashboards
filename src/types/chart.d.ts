export type ChartProps = {
  data: Record<string, string | number>[]
  valueKey: string
  nameKey: string
}
