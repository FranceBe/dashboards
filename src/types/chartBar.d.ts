import any = jasmine.any

export type ChartBarProps = {
  data: Record<string, string | number>[]
  valueKey: string
  nameKey: string
}
