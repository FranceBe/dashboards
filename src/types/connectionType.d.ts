export type ConnectionTypeProps = {
  connection_type: 'ethernet' | 'wifi' | 'cellular'
}
