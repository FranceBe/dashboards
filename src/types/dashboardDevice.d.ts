export type DashboardDeviceable = {
  connection_type: 'ethernet' | 'wifi' | 'cellular'
  last_seen_at: string
  mac_wifi: string
  serial_number: string
  sim_id: number
  status: 'disconnected' | 'connected'
  url: string
  voltage: number
}
