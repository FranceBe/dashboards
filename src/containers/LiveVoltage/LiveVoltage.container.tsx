import Button from 'components/Button'
import ChartLine from 'components/ChartLine'
import {
  ChartAndButtonContainer,
  LiveVoltageContainer,
} from 'containers/LiveVoltage/liveVoltage.style'
import { mockVoltage } from 'containers/LiveVoltage/liveVoltage.utils'
import moment from 'moment'
import React, { useEffect, useState } from 'react'
import { LiveVoltageProps, VoltageChartable } from 'types/liveVoltage'
import { momentFormatter } from 'utils/momentFormatter'

export const LiveVoltage: React.FC<LiveVoltageProps> = ({ deviceId, initialVoltage }) => {
  const [voltagePerTime, setVoltage] = useState<VoltageChartable[]>([initialVoltage])
  const [isLive, setLive] = useState(false)
  useEffect(() => {
    let interval: NodeJS.Timeout
    const getVoltage = () => {
      const currentDate = `${moment(new Date()).utc().format('YYYY-MM-DD HH:mm:ss')}`
      const mockedVoltage = mockVoltage()
      void fetch(`api/device/${deviceId}`)
        // Calling fetch to simulate a realtime call every 5 sec
        // But using mocked data to have a dynamic ChartLine
        .then((res) => res.json())
        .then(() => {
          const newVoltagePerTime = {
            time: momentFormatter(currentDate),
            voltage: mockedVoltage, // here would be the result of the fetch call like "data.voltage"
          }
          setVoltage((prevState: VoltageChartable[]) => [...prevState, newVoltagePerTime])
        })
    }
    if (isLive) {
      getVoltage()
      interval = global.setInterval(() => getVoltage(), 5000)
    }
    return () => {
      clearInterval(interval)
    }
  }, [isLive])

  return (
    <LiveVoltageContainer className={'live-voltage'}>
      <h2> Live voltage stats</h2>
      <ChartAndButtonContainer>
        <ChartLine data={voltagePerTime} valueKey={'voltage'} nameKey={'time'} />
        <Button onClick={() => setLive(!isLive)}>
          {isLive ? 'Stop' : 'Start'} live monitoring
        </Button>
      </ChartAndButtonContainer>
    </LiveVoltageContainer>
  )
}

export default LiveVoltage
