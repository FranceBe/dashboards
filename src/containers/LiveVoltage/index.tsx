import LiveVoltageContainer, { LiveVoltage } from 'containers/LiveVoltage/LiveVoltage.container'

export { LiveVoltage }
export default LiveVoltageContainer
