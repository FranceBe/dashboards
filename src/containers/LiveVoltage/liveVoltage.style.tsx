import styled from 'styled-components'
import { elements, palette, spaces } from 'styles/variables'

export const LiveVoltageContainer = styled.div`
  border: 1px solid ${palette.grey_primary};
  padding: 0 20px 20px 20px;
  border-radius: ${spaces.small};
  box-shadow: ${elements.box_shadow};
  display: flex;
  flex-direction: column;
  h2 {
    text-align: center;
  }
`

export const ChartAndButtonContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  button {
    margin-left: 10%;
    height: 100px;
  }
`
