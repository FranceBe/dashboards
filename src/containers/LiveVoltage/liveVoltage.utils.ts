// This file is meant to mock a voltage variation
// In order to display a live changing ChartLine
export const mockVoltage = (): number => {
  const randomFloat = Math.floor(Math.random() * (1 - 100 + 1)) + 1
  return randomFloat < 0 ? randomFloat * -1 : randomFloat
}
