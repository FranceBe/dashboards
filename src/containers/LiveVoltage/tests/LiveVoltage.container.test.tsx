import '@testing-library/jest-dom/extend-expect'

import { act, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import LiveVoltage from 'containers/LiveVoltage'
import fetch from 'jest-fetch-mock'
import React from 'react'
import { LiveVoltageProps } from 'types/liveVoltage'

describe('LiveVoltage', () => {
  // Set up
  // Mock fetch response
  beforeEach(() => {
    fetch.enableMocks()
    fetch.mockResponseOnce(
      JSON.stringify({
        results: [],
      }),
    )
  })
  const defaultProps: LiveVoltageProps = {
    deviceId: 1,
    initialVoltage: {
      time: 'September 27th 2021, 2:59:09 pm',
      voltage: 12.343,
    },
  }
  it('should match snapshot', async () => {
    const { container } = await waitFor(() => render(<LiveVoltage {...defaultProps} />))
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render title', async () => {
    await waitFor(() => render(<LiveVoltage {...defaultProps} />))
    expect(screen.getByText('Live voltage stats')).toBeInTheDocument()
  })
  it('should render start button', async () => {
    await waitFor(() => render(<LiveVoltage {...defaultProps} />))
    expect(screen.getByText('Start live monitoring')).toBeInTheDocument()
  })
  it('should render stop button when user has clicked', async () => {
    await waitFor(() => render(<LiveVoltage {...defaultProps} />))
    act(() => {
      userEvent.click(screen.getByRole('button'))
    })
    await waitFor(() => expect(screen.getByText('Stop live monitoring')).toBeInTheDocument())
    // Stopping the live monitoring after testing to be clean
    act(() => {
      userEvent.click(screen.getByRole('button'))
    })
  })
  it('should render ChartLine', async () => {
    await waitFor(() => render(<LiveVoltage {...defaultProps} />))
    expect(screen.getByTestId('chart-line')).toBeInTheDocument()
  })
})
