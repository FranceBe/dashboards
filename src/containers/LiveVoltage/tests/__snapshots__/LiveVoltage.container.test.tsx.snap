// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveVoltage should match snapshot 1`] = `
<div
  class="sc-gtsrHT rTluF live-voltage"
>
  <h2>
     Live voltage stats
  </h2>
  <div
    class="sc-dlnjwi jaGdpX"
  >
    <div
      data-testid="chart-line"
    >
      <div
        class="recharts-wrapper"
        style="position: relative; cursor: default; width: 500px; height: 300px;"
      >
        <svg
          class="recharts-surface"
          height="300"
          version="1.1"
          viewBox="0 0 500 300"
          width="500"
        >
          <defs>
            <clippath
              id="recharts1-clip"
            >
              <rect
                height="260"
                width="430"
                x="65"
                y="5"
              />
            </clippath>
          </defs>
          <g
            class="recharts-cartesian-grid"
          >
            <g
              class="recharts-cartesian-grid-horizontal"
            >
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="65"
                x2="495"
                y="5"
                y1="265"
                y2="265"
              />
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="65"
                x2="495"
                y="5"
                y1="200"
                y2="200"
              />
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="65"
                x2="495"
                y="5"
                y1="135"
                y2="135"
              />
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="65"
                x2="495"
                y="5"
                y1="70"
                y2="70"
              />
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="65"
                x2="495"
                y="5"
                y1="5"
                y2="5"
              />
            </g>
            <g
              class="recharts-cartesian-grid-vertical"
            >
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="280"
                x2="280"
                y="5"
                y1="5"
                y2="265"
              />
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="65"
                x2="65"
                y="5"
                y1="5"
                y2="265"
              />
              <line
                fill="none"
                height="260"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="3 3"
                width="430"
                x="65"
                x1="495"
                x2="495"
                y="5"
                y1="5"
                y2="265"
              />
            </g>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
          >
            <line
              class="recharts-cartesian-axis-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              type="category"
              width="430"
              x="65"
              x1="65"
              x2="495"
              y="265"
              y1="265"
              y2="265"
            />
            <g
              class="recharts-cartesian-axis-ticks"
            >
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="30"
                  orientation="bottom"
                  stroke="#666"
                  type="category"
                  width="430"
                  x="65"
                  x1="280"
                  x2="280"
                  y="265"
                  y1="268"
                  y2="265"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="30"
                  orientation="bottom"
                  stroke="none"
                  text-anchor="middle"
                  type="category"
                  width="430"
                  x="280"
                  y="270"
                >
                  <tspan
                    dy="0.71em"
                    x="280"
                  >
                    September 27th 2021, 2:59:09 pm
                  </tspan>
                </text>
              </g>
            </g>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
          >
            <line
              class="recharts-cartesian-axis-line"
              fill="none"
              height="260"
              orientation="left"
              stroke="#666"
              type="number"
              width="60"
              x="5"
              x1="65"
              x2="65"
              y="5"
              y1="5"
              y2="265"
            />
            <g
              class="recharts-cartesian-axis-ticks"
            >
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="260"
                  orientation="left"
                  stroke="#666"
                  type="number"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="265"
                  y2="265"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="260"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  width="60"
                  x="57"
                  y="265"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    0
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="260"
                  orientation="left"
                  stroke="#666"
                  type="number"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="200"
                  y2="200"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="260"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  width="60"
                  x="57"
                  y="200"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    4
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="260"
                  orientation="left"
                  stroke="#666"
                  type="number"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="135"
                  y2="135"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="260"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  width="60"
                  x="57"
                  y="135"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    8
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="260"
                  orientation="left"
                  stroke="#666"
                  type="number"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="70"
                  y2="70"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="260"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  width="60"
                  x="57"
                  y="70"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    12
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="260"
                  orientation="left"
                  stroke="#666"
                  type="number"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="5"
                  y2="5"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="260"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  width="60"
                  x="57"
                  y="5"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    16
                  </tspan>
                </text>
              </g>
            </g>
          </g>
          <g
            class="recharts-layer recharts-line"
          >
            <g
              class="recharts-layer recharts-line-dots"
            >
              <circle
                class="recharts-dot recharts-line-dot"
                cx="280"
                cy="64.42625"
                fill="#fff"
                height="260"
                points="[object Object]"
                r="3"
                stroke="#0E074B"
                stroke-width="1"
                type="monotone"
                width="430"
              />
            </g>
          </g>
        </svg>
        <div
          class="recharts-legend-wrapper"
          style="position: absolute; width: 490px; height: auto; left: 5px; top: 5px;"
        >
          <ul
            class="recharts-default-legend"
            style="padding: 0px; margin: 0px; text-align: center;"
          >
            <li
              class="recharts-legend-item legend-item-0"
              style="display: inline-block; margin-right: 10px;"
            >
              <svg
                class="recharts-surface"
                height="14"
                style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                version="1.1"
                viewBox="0 0 32 32"
                width="14"
              >
                <path
                  class="recharts-legend-icon"
                  d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                  fill="none"
                  stroke="#0E074B"
                  stroke-width="4"
                />
              </svg>
              <span
                class="recharts-legend-item-text"
                style="color: rgb(14, 7, 75);"
              >
                voltage
              </span>
            </li>
          </ul>
        </div>
        <div
          class="recharts-tooltip-wrapper"
          style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
        >
          <div
            class="recharts-default-tooltip"
            style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
          >
            <p
              class="recharts-tooltip-label"
              style="margin: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      class="sc-bdnxRM dcFAlz"
    >
      Start
       live monitoring
    </button>
  </div>
</div>
`;
